pslogin/src/main/scala/WorldSessionActor.scala

```diff
@@ -4163,25 +4104,22 @@ class WorldSessionActor extends Actor
       zoneLoaded = Some(true)
 
     case msg @ PlayerStateMessageUpstream(avatar_guid, pos, vel, yaw, pitch, yaw_upper, seq_time, unk3, is_crouching, is_jumping, jump_thrust, is_cloaking, unk5, unk6) =>
-      if (player.death_by == -1) {
-        sendResponse(ChatMsg(ChatMessageType.UNK_71, true, "", "Your account has been logged out by a Customer Service Representative.", None))
-        Thread.sleep(300)
-        sendResponse(DropSession(sessionId, "kick by GM"))
-      }
+      //log.info(s"$msg")
       turnCounter(avatar_guid)
       val isMoving = WorldEntity.isMoving(vel)
       val isMovingPlus = isMoving || is_jumping || jump_thrust
       if(isMovingPlus) {
         CancelZoningProcessWithDescriptiveReason("cancel_motion")
       }
-
-      if(deadState == DeadState.Alive && upstreamMessageCount % 2 == 0) { // Regen stamina roughly every 500ms
-        if(player.skipStaminaRegenForTurns > 0) {
+      if(deadState == DeadState.Alive) {
+        val regenTick = if(is_crouching) 2 else 3
+        if(player.skipStaminaRegenForTurns > 0 && upstreamMessageCount % 2 == 0) {
           //do not renew stamina for a while
           player.skipStaminaRegenForTurns -= 1
         }
-        else if(!isMovingPlus && player.Stamina != player.MaxStamina) {
-          player.Stamina += 1
+        else if(player.Stamina != player.MaxStamina && !isMovingPlus && upstreamMessageCount % regenTick == 0) {
```

Comment:
From previous comment:

>Where has this come from? All rexo footage we found from live showed a consistent 500ms stamina tick, no matter what stance the player is in.

Reply:
Wiki.
That's also code from before ... well, long before.  You guys had that video dive session after that mess with the first time we merged inventory things.  I don't know if you ever got back to me with stamina-related answers.